caffe2/core/predictor.cc

```diff
@@ -51,80 +33,121 @@ const ::google::protobuf::RepeatedPtrField<::std::string>& getBlobs(
   }
   CAFFE_THROW("Blob not found: ", name);
 }
-} // namespace
 
-Predictor::Predictor(const MetaNetDef& def, Workspace* parent)
-    : Predictor(
-          getNet(
+void shareInputTensor(
+    Workspace* ws,
+    const std::string& name,
+    TensorCPU* input) {
+  enforceIsTensor<TensorCPU>(ws, name);
+  auto* blob = ws->GetBlob(name);
+  CAFFE_ENFORCE(blob, "Blob: ", name, " does not exist");
+  auto* tensor = blob->template GetMutable<TensorCPU>();
+  tensor->ResizeLike(*input);
+  tensor->ShareData(*input);
+}
+
+std::shared_ptr<TensorCPU> extractOutputTensor(Workspace* ws, const std::string& name) {
+  enforceIsTensor<TensorCPU>(ws, name);
+  auto* blob = ws->GetBlob(name);
+  CAFFE_ENFORCE(blob, "Blob: ", name, " does not exist");
+
+  // Since the tensor is a member of blob, we should not delete the pointer.
+  // By given a customized deleter, the pointer won't be deleted on shared_ptr deconstruction.
+  return std::shared_ptr<TensorCPU>(blob->template GetMutable<TensorCPU>(), [](TensorCPU *){});
```

Comment:
This is hacky. Why do we need to change the interface to return `std::shared_ptr<TensorCPU>` in the first place? I feel that doing this defeats the purpose of `shared_ptr`.

Reply:
The reason of changing the interface from raw pointer to `shared_ptr` is to make it compatible to thread-safe evaluation case, in which we temporarily create a new `Workspace` and new `Blob`'s for calculation. Therefore, we should take responsibility to deallocate the memory allocation we have made for those temporary `Tensor`'s, and `shared_ptr` is the only clean solution. Then for non-thread-safe case, in order to make them share the same interface, we have to do these hacks.